File: Anime Game Remap (for all users)/api/src/cpp/core/src/model/strategies/ModType.cpp
#include "AGRemapCore/model/strategies/ModType.h"

#include <set>
#include <utility>

#include "AGRemapCore/constants/IniKeywords.h"
#include "AGRemapCore/model/files/IniFile.h"
#include "AGRemapCore/tools/Heading.h"
#include "AGRemapCore/tools/StringTools.h"

#include "AGRemapCore/data/ModDataAssets.h"

#include "AGRemapCore/constants/GlobalIniRemoveBuilders.h"


namespace AGRemapCore {
    ModType::ModType(int gameTypeId, int modTypeId, const std::string &name, const std::vector<std::string> &aliases,
                      std::shared_ptr<Hashes> hashes, std::shared_ptr<Indices> indices,
                      std::shared_ptr<VertexCounts> vertexCounts, std::shared_ptr<VGRemaps> vgRemaps,
                      std::shared_ptr<IniParseBuilder> iniParseBuilder, std::shared_ptr<IniFixBuilder> iniFixBuilder,
                      std::shared_ptr<IniRemoveBuilder> iniRemoveBuilder):
        gameTypeId(gameTypeId), modTypeId(modTypeId), name(name), aliases(aliases),
        hashes(std::move(hashes)), indices(std::move(indices)), vertexCounts(std::move(vertexCounts)),
        vgRemaps(std::move(vgRemaps)),
        iniParseBuilder(std::move(iniParseBuilder)), iniFixBuilder(std::move(iniFixBuilder)), iniRemoveBuilder(std::move(iniRemoveBuilder)) {
        // Mirrors the pure-Python original's own "if (hashes is None): hashes = Hashes()". Note a
        // default-constructed Hashes is fully populated, so this is a real table, not an empty one.
        // A fresh instance per ModType, exactly as the original does -- see ModType::hashes on why
        // that (rather than one shared table) is the right default.
        if (this->hashes == nullptr) {
            this->hashes = std::make_shared<Hashes>();
        }

        // The same, for the original's "if (indices is None): indices = Indices()".
        if (this->indices == nullptr) {
            this->indices = std::make_shared<Indices>();
        }

        // ...and for "if (vertexCounts is None): vertexCounts = VertexCounts()".
        if (this->vertexCounts == nullptr) {
            this->vertexCounts = std::make_shared<VertexCounts>();
        }

        // vgRemaps is the odd one out: the original falls back to the SHARED
        // "ModDataAssets.VGRemaps.value", not to a fresh VGRemaps(). Mirrored deliberately -- see
        // ModType::vgRemaps' warning about the mutation consequences.
        if (this->vgRemaps == nullptr) {
            this->vgRemaps = ModDataAssets::vgRemaps();
        }

        // Mirrors the pure-Python original's own "if (iniParseBuilder is None): iniParseBuilder =
        // IniParseBuilder(GIMIParser)" fallback -- a default-constructed IniParseBuilder builds a
        // plain BaseIniParser, since no concrete GIMIParser equivalent exists in C++ yet (see
        // IniParseBuilder::defaultFactory, the single place to change when one lands).
        if (this->iniParseBuilder == nullptr) {
            this->iniParseBuilder = std::make_shared<IniParseBuilder>();
        }

        // Same fallback on the fixer side, mirroring the original's own "iniFixBuilder" default.
        if (this->iniFixBuilder == nullptr) {
            this->iniFixBuilder = std::make_shared<IniFixBuilder>();
        }

        // Mirrors the pure-Python original's own
        // "iniRemoveBuilder = GlobalIniRemoveBuilders.RemoveBuilder.value" default. Note this is
        // the *shared* global builder rather than a fresh one, so every ModType that falls back
        // here shares one builder instance -- deliberate, and what the original does too.
        if (this->iniRemoveBuilder == nullptr) {
            this->iniRemoveBuilder = GlobalIniRemoveBuilders::removeBuilder();
        }
    }


    bool ModType::isName(const std::string& name) const {
        // Case-insensitive by grapheme through StringTools (the pure-Python original's own
        // '.lower()' comparison on a Unicode str), so a non-ASCII name or alias compares correctly.
        if (StringTools::equalsIgnoreCase(this->name, name)) {
            return true;
        }

        for (const std::string& alias : aliases) {
            if (StringTools::equalsIgnoreCase(alias, name)) {
                return true;
            }
        }

        return false;
    }

    std::unordered_set<std::string> ModType::getModsToFix() const {
        // See the header's warning: the pure-Python original reads two never-populated 'fixTo'
        // sets and so always answers "none". These are the targets the remap tables actually hold,
        // keyed by this mod type's own name.
        std::unordered_set<std::string> result;

        auto addFrom = [this, &result](const auto& assets) {
            if (assets == nullptr) {
                return;
            }

            const auto& map = assets->getMap();
            auto it = map.find(name);
            if (it == map.end()) {
                return;
            }

            result.insert(it->second.begin(), it->second.end());
        };

        addFrom(hashes);
        addFrom(indices);

        return result;
    }

    std::optional<int> ModType::getVertexCount(const std::optional<Version>& version) const {
        if (vertexCounts == nullptr) {
            return std::nullopt;
        }

        // Two non-version columns, name then component -- VertexCounts' own class note. The
        // pure-Python original passes only the name because its ModAssets.get pads; ModDictAssets
        // requires both, and "" is the component every shipped row carries (a real key value, not
        // a "missing" marker), so it is what a caller wanting a mod's overall count passes.
        //
        // errorOnNotFound = false: the pure-Python original leaves its own default in place and so
        // raises for a mod type with no row, which is not a useful way to say "unknown" in C++.
        return vertexCounts->get({name, ""}, version, false);
    }

    std::optional<VGRemap> ModType::getVGRemap(const std::string& modName, const std::optional<Version>& fromVersion,
                                                const std::optional<Version>& toVersion,
                                                const std::optional<std::string>& fromComp,
                                                const std::optional<std::string>& toComp) const {
        if (vgRemaps == nullptr) {
            return std::nullopt;
        }

        // VGRemaps' four non-version columns are, in order: fromChar, fromComp, toChar, toComp
        // (see its own class note). A std::nullopt leaves that column unconstrained, which is what
        // the pure-Python original expresses by simply leaving the key out of its dict.
        return vgRemaps->get({name, fromComp, modName, toComp}, {fromVersion, toVersion}, false);
    }

    std::string ModType::getHelpStr() const {
        Heading modTypeHeading(name, 8, "-");

        std::string result = modTypeHeading.open();
        result += "\n\nname: " + name;

        if (!aliases.empty()) {
            // Sorted, matching the original -- the help text is user-facing, so a stable order
            // matters more than the order the aliases happen to be declared in.
            std::set<std::string> sortedAliases(aliases.begin(), aliases.end());

            std::string aliasStr;
            for (const std::string& alias : sortedAliases) {
                if (!aliasStr.empty()) {
                    aliasStr += ", ";
                }
                aliasStr += alias;
            }

            result += "\naliases: " + aliasStr;
        }

        result += "\n\n" + modTypeHeading.close();
        return result;
    }

    void ModType::fixIni(IniFile& iniFile, bool keepBackup, bool fixOnly) const {
        const ModType* iniModType = iniFile.getAvailableType();
        if (iniModType == nullptr || iniModType->name != name) {
            return;
        }

        iniFile.fix(keepBackup, fixOnly);
    }

    Ranges<long long> ModType::getHashRanges(const IfContentPartColouring<std::string, std::string>& partColours,
                                              const std::optional<Version>& version,
                                              const std::vector<std::optional<std::string>>& nonVersionVals) const {
        std::unordered_map<std::string, bool> keysExists;
        keysExists.emplace(IniKeywords::Hash, true);

        std::unordered_map<std::string, IfContentPartColouring<std::string, std::string>::Filter> keyFilters;
        keyFilters.emplace(IniKeywords::Hash,
                           [this, &version, &nonVersionVals](std::optional<long long>, const std::string& value) {
            return hashes != nullptr && hashes->hasFrom(value, version, nonVersionVals);
        });

        return partColours.getRanges(keysExists, keyFilters);
    }
}

File: Anime Game Remap (for all users)/api/src/py/FixRaidenBoss2/model/strategies/ModType.py
##### ExtImports
from typing import Union, Optional, List, Set, TYPE_CHECKING, Hashable, Dict, Type
##### EndExtImports

##### CppLocalImports
from ...core import Ranges
from ...core import IfContentPartColouring
from ...core import Hashes
from ...core import Indices
from ...core import VGRemap
##### EndCppLocalImports

##### LocalImports
from ...constants.GlobalIniRemoveBuilders import GlobalIniRemoveBuilders
from ...constants.GenericTypes import VersionType
from ...constants.IniConsts import IniKeywords
from ..assets.VertexCounts import VertexCounts
from ..assets.VGRemaps import VGRemaps
from ...tools.ListTools import ListTools
from ...tools.DictTools import UnHashableNone
from ...tools.Heading import Heading
from ...model.strategies.iniParsers.IniParseBuilder import IniParseBuilder
from ...core import GIMIParser
from ...model.strategies.iniFixers.IniFixBuilder import IniFixBuilder
from ...model.strategies.iniFixers.GIMIFixerOld import GIMIFixerOld
from ...model.strategies.iniRemovers.IniRemoveBuilder import IniRemoveBuilder
from ...data.ModDataAssets import ModDataAssets

if (TYPE_CHECKING):
    from ..files.IniFile import IniFile
##### EndLocalImports


##### Script
class ModType():
    """
    Class for defining a generic type of mod

    Parameters
    ----------
    name: :class:`str`
        The default name for the type of mod

    hashes: Optional[:class:`Hashes`]
        The hashes related to the mod and its fix :raw-html:`<br />` :raw-html:`<br />`

        If this value is ``None``, then will create a new, empty :class:`Hashes` :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    indices: Optional[:class:`Indices`]
        The indices related to the mod and its fix :raw-html:`<br />` :raw-html:`<br />`

        If this ``None``, then will create a new, emtpy :class:`Indices` :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    aliases: Optional[List[:class:`str`]]
        Other alternative names for the type of mod :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    vgRemaps: Optional[:class:`VGRemaps`]
        Maps the blend indices from the vertex group of one mod to another mod :raw-html:`<br />` :raw-html:`<br />`

        If this value is ``None``, then will use the global predefined :class:`VGRemaps` at :class:`ModDataAssets` :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    iniParseBuilder: Optional[:class:`IniParseBuilder`]
        The builder to build the parser used for .ini files :raw-html:`<br />` :raw-html:`<br />`

        If this value is ``None``, then by default this attribute will be set to **IniParseBuilder(:class:`GIMIParser`)** :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    iniFixBuilder: Optional[:class:`IniFixBuilder`]
        The builder to build the fixer used for .ini files :raw-html:`<br />` :raw-html:`<br />`

        If this value is ``None``, then by default this attribute will be set to **IniFixBuilder(:class:`GIMIFixer`)** :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    iniRemoveBuilder: Optional[:class:`IniRemoveBuilder`]
        The builder to build the remover used for .ini files :raw-html:`<br />` :raw-html:`<br />`

        If this value is ``None``, then by default this attribute will be set to **IniRemoveBuilder(:class:`IniRemover`)** :raw-html:`<br />` :raw-html:`<br />`

        **Default**: ``None``

    Attributes
    ----------
    name: :class:`str`
        The default name for the type of mod

    hashes: :class:`Hashes`
        The hashes related to the mod and its fix

    indices: :class:`Indices`
        The indices related to the mod and its fix

    vertexCounts: :class:`VertexCounts`
        The vertex counts related to the mod and its fix

    vgRemaps: :class:`VGRemaps`
        The repository that stores the mapping for remapping vertex group blend indices of the mod to the vertex group blend indices of another mod

    aliases: Optional[List[:class:`str`]]
        Other alternative names for the type of mod

    iniParseBuilder: :class:`IniParseBuilder`
        The builder to build the parser used for .ini files

    iniFixBuilder: :class:`IniFixBuilder`
        the builder to build the fixer used for .ini files

    iniRemoveBuilder: :class:`IniRemoveBuilder`
        the builder to build the remover used for .ini files
    """

    def __init__(self, name: str, hashes: Optional[Hashes] = None, indices: Optional[Indices] = None, vertexCounts: Optional[VertexCounts] = None,
                 aliases: Optional[List[str]] = None, vgRemaps: Optional[VGRemaps] = None,
                 iniParseBuilder: Optional[IniParseBuilder] = None, iniFixBuilder: Optional[IniFixBuilder] = None, iniRemoveBuilder: Optional[IniRemoveBuilder] = None):
        self.name = name
        if (hashes is None):
            hashes = Hashes()

        if (indices is None):
            indices = Indices()

        if (vertexCounts is None):
            vertexCounts = VertexCounts()

        self.hashes = hashes
        self.indices = indices
        self.vertexCounts = vertexCounts
        
        if (aliases is None):
            aliases = []
        self.aliases = ListTools.getDistinct(aliases)
        
        self._maxVgIndex = None
        if (vgRemaps is None):
            vgRemaps = ModDataAssets.VGRemaps.value

        self.vgRemaps = vgRemaps

        if (iniParseBuilder is None):
            iniParseBuilder = IniParseBuilder(GIMIParser)

        if (iniFixBuilder is None):
            iniFixBuilder = IniFixBuilder(GIMIFixerOld)

        if (iniRemoveBuilder is None):
            iniRemoveBuilder = GlobalIniRemoveBuilders.RemoveBuilder.value

        self.iniParseBuilder = iniParseBuilder
        self.iniFixBuilder = iniFixBuilder
        self.iniRemoveBuilder = iniRemoveBuilder

    def isName(self, name: str) -> bool:
        """
        Determines whether a certain name matches with the names defined for this type of mod

        Parameters
        ----------
        name: :class:`str`
            The name being searched

        Returns
        -------
        :class:`bool`
            Whether the searched name matches with the names for this type of mod
        """

        name = name.lower()
        if (self.name.lower() == name):
            return True
        
        for alias in self.aliases:
            if (alias.lower() == name):
                return True

        return False

    def getModsToFix(self) -> Set[str]:
        """
        Retrieves the names of the mods this mod type will fix to

        Returns
        -------
        Set[:class:`str`]
            The names of the mods to fix to
        """

        result = set()
        result = result.union(self.hashes.fixTo)
        result = result.union(self.indices.fixTo)
        return result
    
    def getVertexCount(self, version: Optional[Union[str, float, VersionType]] = None) -> int:
        """
        Retrieves the number of vertices for a mod

        .. attention::
            This function assumes that the specified dictionary :attr:`vertexCounts` (:attr:`VertexCounts.map`) contains :attr:`name` (the name of this mod type) as a mod to get the vertex count for

        Parameters
        ----------
        version: Optional[Union[:class:`str`, :class:`float`, `packaging.version.Version`_]]
            The specific game version we want for the vertex count :raw-html:`<br />` :raw-html:`<br />`

            If this value is ``None``, then will get the latest version of the vertex count :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        Returns 
        -------
        :class:`int`
            The number of vertices for the mod
        """

        return self.vertexCounts.get(self.name, versionVals = version)
    
    def getVGRemap(self, modName: str, fromVersion: Optional[Union[str, float, VersionType]] = None, toVersion: Optional[Union[str, float, VersionType]] = None, fromComp: Optional[str] = None, toComp: Optional[str] = None) -> VGRemap:
        """
        Retrieves the corresponding Vertex Group Remap

        .. attention::
            This function assumes that the specified map :attr:`vgRemaps` (:attr:`VGRemaps.map`) contains :attr:`name` (the name of this mod type) as a mod to map from

        Parameters
        ----------
        modName: :class:`str`
            The name of the mod to map to

        fromVersion: Optional[Union[:class:`str`, :class:`float`, `packaging.version.Version`_]]
            The specific game version we want for to remap from :raw-html:`<br />` :raw-html:`<br />`

            If this value is ``None``, then to remap from the latest version of the original mod :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        toVersion: Optional[Union[:class:`str`, :class:`float`, `packaging.version.Version`_]]
            The specific game version we want for to remap to :raw-html:`<br />` :raw-html:`<br />`

            If this value is ``None``, then to remap to the latest version of the new mod :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        fromComp: Optional[:class:`str`]
            The specific component to remap from in the original mod :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        toComp: Optional[:class:`str`]
            The specific component to remap to in the new mod :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        Returns 
        -------
        :class:`VGRemap`
            The corresponding remap
        """

        versionVals = {"fromVersion": fromVersion, "toVersion": toVersion}
        nonVersionVals = {"fromChar": self.name, "toChar": modName}

        if (fromComp is not None):
            nonVersionVals["fromComp"] = fromComp

        if (toComp is not None):
            nonVersionVals["toComp"] = toComp

        return self.vgRemaps.get(nonVersionVals = nonVersionVals, versionVals = versionVals)

    def getHelpStr(self) -> str:
        modTypeHeading = Heading(self.name, 8, "-")

        currentHelpStr = f"{modTypeHeading.open()}"
        currentHelpStr += f"\n\nname: {self.name}"
        
        if (self.aliases):
            sortedAliases = sorted(self.aliases)
            aliasStr = ", ".join(sortedAliases)
            currentHelpStr += f"\naliases: {aliasStr}"

        currentHelpStr += f"\n\n{modTypeHeading.close()}"
        return currentHelpStr
    
    def fixIni(self, iniFile: "IniFile", keepBackup: bool = True, fixOnly: bool = False):
        """
        Fixes the .ini file associated to this type of mod

        Parameters
        ----------
        iniFile: :class:`IniFile`
            The .ini file to fix

        keepBackup: :class:`bool`
            Whether to keep backups for the .ini file :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``True``

        fixOnly: :class:`bool`
            Whether to only fix the .ini file without undoing any fixes :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``False``
        """

        iniModType = iniFile.availableType
        if (iniModType is not None and iniModType.name == self.name):
            iniFile.fix(keepBackup = keepBackup, fixOnly = fixOnly)

    def getHashRanges(self, partColours: IfContentPartColouring, version: Optional[Union[str, float, VersionType]] = None, nonVersionVals: Optional[Union[Hashable, List[Hashable], Dict[str, Hashable], Type[UnHashableNone]]] = UnHashableNone) -> Ranges:
        """
        Retrieves the valid ranges of order indices for some :class:`IfContentPart` based on the hashes of the mod

        Parameters
        ----------
        partColours: :class:`IfContentPartColouring`
            The current states of the :class:`IfContentPart`

        version: Optional[Union[:class:`float`, :class:`str`, `packaging.version.Version`_]]
            The version we want the hashes to come from, See :meth:`Hashes.hasFrom` for details :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        nonVersionVals: Optional[Union[`Hashable`_, List[`Hashable`_], Dict[:class:`str`, `Hashable`_], Type[:class:`UnHashableNone`]]]
            The values to the non-version indices used to help filter for finding a particular instance of some hash. See :meth:`Hashes.hasFrom` for details. :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        Returns
        -------
        :class:`Ranges`
            The valid ranges of indices for the :class:`IfContentPart`
        """

        return partColours.getRanges(keysExists = {IniKeywords.Hash.value: True}, keyFilters = {IniKeywords.Hash.value: lambda ind, val: self.hashes.hasFrom(val, version = version, nonVersionVals = nonVersionVals)})
##### EndScript